Preprocessor memory leak - unterminated function-like macro invocation

This only manifests on Linux, I wasn't able to reproduce it on MacOS. Zig version `0.9.0-dev.1561+5ebdc8c46` 

```C
#define foo(X) X
foo(1,
```
```
root@006bbb66b04e:/arocc# ./zig-out/bin/arocc -E test/cases/leak.c
test/cases/leak.c:2:1: error: unterminated function macro argument list
foo(1,
^
1 error generated.
foo ( 1 ,
error(gpa): memory address 0x7fd5a77e0380 leaked:
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:340:69: 0x349579 in std.array_list.ArrayListAligned([]const Tree.Token,null).ensureTotalCapacityPrecise (arocc)
                const new_memory = try self.allocator.reallocAtLeast(self.allocatedSlice(), new_capacity);
                                                                    ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:325:55: 0x27babe in std.array_list.ArrayListAligned([]const Tree.Token,null).ensureTotalCapacity (arocc)
                return self.ensureTotalCapacityPrecise(better_capacity);
                                                      ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:364:41: 0x376b78 in std.array_list.ArrayListAligned([]const Tree.Token,null).addOne (arocc)
            try self.ensureTotalCapacity(newlen);
                                        ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:175:49: 0x3491c6 in std.array_list.ArrayListAligned([]const Tree.Token,null).append (arocc)
            const new_item_ptr = try self.addOne();
                                                ^


error(gpa): memory address 0x7fd5a77e2500 leaked:
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:340:69: 0x321489 in std.array_list.ArrayListAligned(Tree.Token,null).ensureTotalCapacityPrecise (arocc)
                const new_memory = try self.allocator.reallocAtLeast(self.allocatedSlice(), new_capacity);
                                                                    ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:325:55: 0x27bbde in std.array_list.ArrayListAligned(Tree.Token,null).ensureTotalCapacity (arocc)
                return self.ensureTotalCapacityPrecise(better_capacity);
                                                      ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:364:41: 0x321328 in std.array_list.ArrayListAligned(Tree.Token,null).addOne (arocc)
            try self.ensureTotalCapacity(newlen);
                                        ^
/zig-linux-x86_64-0.9.0-dev.1561+5ebdc8c46/lib/std/array_list.zig:175:49: 0x2fa4c6 in std.array_list.ArrayListAligned(Tree.Token,null).append (arocc)
            const new_item_ptr = try self.addOne();
                                                ^
```

--- a/src/Preprocessor.zig
+++ b/src/Preprocessor.zig
@@ -550,6 +550,12 @@ fn skipToNl(tokenizer: *Tokenizer) void {
 
 const ExpandBuf = std.ArrayList(Token);
 const MacroArguments = std.ArrayList([]const Token);
+fn deinitMacroArguments(allocator: *Allocator, args: *const MacroArguments) void {
+    for (args.items) |item| {
+        allocator.free(item);
+    }
+    args.deinit();
+}
 
 fn expandObjMacro(pp: *Preprocessor, simple_macro: *const Macro) Error!ExpandBuf {
     var buf = ExpandBuf.init(pp.comp.gpa);
@@ -847,7 +853,7 @@ fn collectMacroFuncArguments(
     // collect the arguments.
     var parens: u32 = 0;
     var args = MacroArguments.init(pp.comp.gpa);
-    errdefer args.deinit();
+    errdefer deinitMacroArguments(pp.comp.gpa, &args);
     var curArgument = std.ArrayList(Token).init(pp.comp.gpa);
     defer curArgument.deinit();
     var done = false;
@@ -875,6 +881,7 @@ fn collectMacroFuncArguments(
                 }
             },
             .eof => {
+                deinitMacroArguments(pp.comp.gpa, &args);
                 tokenizer.index = initial_tokenizer_index;
                 end_idx.* = old_end;
                 try pp.comp.addDiagnostic(.{ .tag = .unterminated_macro_arg_list, .loc = name_tok.loc });
@@ -953,7 +960,7 @@ fn expandMacroExhaustive(pp: *Preprocessor, tokenizer: *Tokenizer, buf: *ExpandB
                         continue;
                     }
                     var expanded_args = MacroArguments.init(pp.comp.gpa);
-                    defer expanded_args.deinit();
+                    defer deinitMacroArguments(pp.comp.gpa, &expanded_args);
                     try expanded_args.ensureCapacity(args.items.len);
                     for (args.items) |arg| {
                         var expand_buf = ExpandBuf.init(pp.comp.gpa);
@@ -966,9 +973,6 @@ fn expandMacroExhaustive(pp: *Preprocessor, tokenizer: *Tokenizer, buf: *ExpandB
 
                     var res = try pp.expandFuncMacro(macro_tok.loc, macro, &args, &expanded_args);
                     defer res.deinit();
-                    for (expanded_args.items) |arg| {
-                        pp.comp.gpa.free(arg);
-                    }
                     var expansion_loc = macro.loc;
                     expansion_loc.next = buf.items[idx].loc.next;
                     for (res.items) |*tok| {

--- a/test/cases/unterminated macro function at eof.c
+++ b/test/cases/unterminated macro function at eof.c
@@ -1,0 +1,9 @@
+#define EXPECTED_ERRORS "unterminated macro function at eof.c:8:1: error: unterminated function macro argument list" \
+	"unterminated macro function at eof.c:8:1: warning: type specifier missing, defaults to 'int'" \
+	"unterminated macro function at eof.c:8:5: error: expected parameter declaration" \
+	"unterminated macro function at eof.c:8:5: error: expected ')', found 'an integer literal'" \
+	"unterminated macro function at eof.c:8:4: note: to match this '('"
+
+#define foo(X) X
+foo(1,
+

